src/arch/riscv/machine/hardware.c

```diff
@@ -62,67 +62,70 @@ BOOT_CODE void map_kernel_devices(void)
  * are for user level, and also call plic_complete_claim for seL4_IRQHandler_Ack.
  */
 
-/**
- * Gets the new active irq from the PLIC or STIP.
- *
- * getNewActiveIRQ is only called by getActiveIRQ and checks for a pending IRQ.
- * We read sip and if the SEIP bit is set we claim an
- * IRQ from the PLIC. If STIP is set then it is a kernel timer interrupt.
- * Otherwise we return IRQ invalid. It is possible to reveive irqInvalid from
- * the PLIC if another HART context has claimed the IRQ before us. This function
- * is not idempotent as plic_get_claim is called which accepts an IRQ message
- * from the PLIC and will claim different IRQs if called subsequent times.
- *
- * @return     The new active irq.
- */
-static irq_t getNewActiveIRQ(void)
-{
-
-    uint64_t sip = read_sip();
-    /* Interrupt priority (high to low ): external -> software -> timer */
-    if (sip & BIT(SIP_SEIP)) {
-        return plic_get_claim();
-#ifdef ENABLE_SMP_SUPPORT
-    } else if (sip & BIT(SIP_SEIP)) {
-        sbi_clear_ipi();
-        return ipi_get_irq();
-#endif
-    } else if (sip & BIT(SIP_STIP)) {
-        return KERNEL_TIMER_IRQ;
-    }
-
-    return irqInvalid;
-}
-
 static irq_t active_irq[CONFIG_MAX_NUM_NODES];
 
 
 /**
  * Gets the active irq. Returns the same irq if called again before ackInterrupt.
  *
- * getActiveIRQ is used to return a currently pending IRQ. This function can be
- * called multiple times and needs to return the same IRQ until ackInterrupt is
- * called. getActiveIRQ returns irqInvalid if no interrupt is pending. It is
- * assumed that if isIRQPending is true, then getActiveIRQ will not return
- * irqInvalid. getActiveIRQ will call getNewActiveIRQ and cache its result until
- * ackInterrupt is called.
+ * This function is called by the kernel to get the interrupt that is currently
+ * active. It not interrupt is currently active, it will try to find one and
+ * put it in the active state. If no interrupt is found, irqInvalid is returned.
+ * It can't be assumed that if isIRQPending() returned true, there will always
+ * be an active interrupt then this is called. It may hold in mayn cases, but
+ * there are corner cases with level-triggered interrupts or on multicore
+ * systems.
+ * This function can be called multiple times during one kernel entry. It must
+ * guarantee that once one interrupt is reported as active, this interrupt is
+ * always returned until ackInterrupt() is called eventually.
  *
- * @return     The active irq.
+ * @return     The active irq or irqInvalid.
  */
 static inline irq_t getActiveIRQ(void)
 {
+    irq_t *active_irq_slot = &active_irq[CURRENT_CPU_INDEX()];
 
-    uint32_t irq;
-    if (!IS_IRQ_VALID(active_irq[CURRENT_CPU_INDEX()])) {
-        active_irq[CURRENT_CPU_INDEX()] = getNewActiveIRQ();
```

Comment:
It was originally done this way in order to be fairly identical to the Arm verified platform: https://github.com/seL4/seL4/blob/master/include/arch/arm/arch/machine/gic_v2.h#L162 to make verification easier and for implementation consistency.

Reply:
I can rework it back and postpone this change if that is a blocker.

Reply:
Up to @lsf37 as I only have a weak opinion that the implementation should follow a similar pattern across architectures if possible, but it doesn't have to.